man/rcvtty.man
.TH RCVTTY %manext1% 1999-04-30 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
rcvtty \- report new mail with nmh
.SH SYNOPSIS
.HP 5
.na
.B %nmhlibexecdir%/rcvtty
.RB [ \-help ]
.RB [ \-version ]
.RI [ command ]
.RB [ \-form
.IR formatfile ]
.RB [ \-format
.IR string ]
.RB [ \-width
.IR columns ]
.RB [ \-bell " | " \-nobell ]
.RB [ \-newline " | " \-nonewline ]
.RB [ \-biff ]
.ad
.SH DESCRIPTION
The
.B rcvtty
command can be used to report new mail.  It is used primarily
in conjunction with mail filtering agents such as
.B slocal
or
.BR procmail .
.PP
.B rcvtty
executes the named command with the message as its standard input,
and writes the resulting output on your terminal.
.PP
Alternatively, if no command is specified (or the command is bogus), then
.B rcvtty
will instead write a one-line scan listing.  The default output format
of this scan listing may be overridden by using either the
.B \-form
.I formatfile
or
.B \-format
.I string
option, similar to the equivalent options for
.B scan
and
.BR inc .
See
.IR mh\-format (5)
for details.
.PP
A newline is output before the message output, and the terminal bell is
rung after the output.  The
.B \-nonewline
and
.B \-nobell
switches will inhibit these functions.
.PP
The switch
.B \-width
.I columns
may be given to specify the width of the scan line.
The default is to use the width of the terminal.
.PP
In addition to the standard format escapes described in
.IR mh\-format (5),
.B rcvtty
also recognizes the following additional
.I component
escapes:
.PP
.RS 5
.nf
.ta \w'Dtimenow  'u +\w'Returns  'u
Escape	Returns	Description
body	string	the (compressed) first part of the body
dtimenow	date	the current date
folder	string	the name of the current folder
.fi
.RE
.PP
By default,
.B rcvtty
will send its output to every terminal on the local machine that
is owned by current user, and that has given write permission
as granted by the command
.IR mesg (1).
If the option
.B \-biff
is given, then
.B rcvtty
will obey the notification status set by the command
.IR biff (1)
instead.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^%nmhetcdir%/mts.conf~^nmh mts configuration file
^$HOME/.maildelivery~^The file controlling local delivery
^%nmhetcdir%/maildelivery~^Rather than the standard file
.fi
.SH CONTEXT
None
.SH DEFAULTS
.nf
.RB ` \-width "' defaults to the width of the terminal"
.RB ` \-newline '
.RB ` \-bell '
.fi
.SH "SEE ALSO"
.IR mh-format (5),
.IR rcvdist (1),
.IR rcvpack (1),
.IR rcvstore (1),
.IR slocal (1)
.SH BUGS
Only two return codes are meaningful, others should be.


NEWS
#
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.
- mhfixmsg now adds a Content-Transfer-Encoding header at the message level,
  if needed after decoding text parts.

uip/rcvtty.c
/* rcvtty.c -- a rcvmail program (a lot like rcvalert) handling IPC ttys
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

/* Changed to use getutent() and friends.  Assumes that when getutent() exists,
 * a number of other things also exist.  Please check.
 * Ruud de Rooij <ruud@ruud.org>  Sun, 28 May 2000 17:28:55 +0200
 */

#include <h/mh.h>
#include <h/signals.h>
#include <setjmp.h>
#include <h/rcvmail.h>
#include <h/scansbr.h>
#include <h/tws.h>
#include <h/mts.h>
#include <h/utils.h>
#include "../sbr/m_mktemp.h"
#include <fcntl.h>

#ifdef HAVE_GETUTXENT
#include <utmpx.h>
#endif /* HAVE_GETUTXENT */

#define	SCANFMT	\
"%2(hour{dtimenow}):%02(min{dtimenow}): %<(size)%5(size) %>%<{encrypted}E%>\
%<(mymbox{from})%<{to}To:%14(friendly{to})%>%>%<(zero)%17(friendly{from})%>  \
%{subject}%<{body}<<%{body}>>%>"

#define RCVTTY_SWITCHES \
    X("biff", 0, BIFFSW) \
    X("form formatfile", 0, FORMSW) \
    X("format string", 5, FMTSW) \
    X("width columns", 0, WIDTHSW) \
    X("newline", 0, NLSW) \
    X("nonewline", 0, NNLSW) \
    X("bell", 0, BELSW) \
    X("nobell", 0, NBELSW) \
    X("version", 0, VERSIONSW) \
    X("help", 0, HELPSW) \

#define X(sw, minchars, id) id,
DEFINE_SWITCH_ENUM(RCVTTY);
#undef X

#define X(sw, minchars, id) { sw, minchars, id },
DEFINE_SWITCH_ARRAY(RCVTTY, switches);
#undef X

static jmp_buf myctx;
static int bell = 1;
static int newline = 1;
static int biff = 0;
static int width = -1;
static char *form = NULL;
static char *format = NULL;

/*
 * external prototypes
 */
char *getusername(void);

/*
 * static prototypes
 */
static void alrmser (int);
static int message_fd (char **);
static int header_fd (void);
#if HAVE_GETUTXENT
static void alert (char *, int);
#endif /* HAVE_GETUTXENT */


int
main (int argc, char **argv)
{
    int md, vecp = 0;
    char *cp, *user, buf[BUFSIZ], tty[BUFSIZ];
    char **argp, **arguments, *vec[MAXARGS];
    struct utmpx *utp;

    if (nmh_init(argv[0], 2)) { return 1; }

    mts_init ();
    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW:
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW:
		    vec[vecp++] = --cp;
		    continue;

		case HELPSW:
		    snprintf (buf, sizeof(buf), "%s [command ...]", invo_name);
		    print_help (buf, switches, 1);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case BIFFSW:
		    biff = 1;
		    continue;

		case FORMSW:
		    if (!(form = *argp++) || *form == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    format = NULL;
		    continue;
		case FMTSW:
		    if (!(format = *argp++) || *format == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    form = NULL;
		    continue;

		case WIDTHSW:
		    if (!(cp = *argp++) || *cp == '-')
			adios(NULL, "missing argument to %s", argp[-2]);
		    width = atoi(cp);
		    continue;
                case NLSW:
                    newline = 1;
                    continue;
                case NNLSW:
                    newline = 0;
                    continue;
                case BELSW:
                    bell = 1;
                    continue;
                case NBELSW:
                    bell = 0;
                    continue;

	    }
	}
	vec[vecp++] = cp;
    }
    vec[vecp] = 0;

    if ((md = vecp ? message_fd (vec) : header_fd ()) == NOTOK)
	exit (RCV_MBX);

    user = getusername();

#if HAVE_GETUTXENT
    setutxent();
    while ((utp = getutxent()) != NULL) {
        if (utp->ut_type == USER_PROCESS && utp->ut_user[0] != 0
               && utp->ut_line[0] != 0
               && strncmp (user, utp->ut_user, sizeof(utp->ut_user)) == 0) {
            strncpy (tty, utp->ut_line, sizeof(utp->ut_line));
	    alert (tty, md);
        }
    }
    endutxent();
#else
    NMH_UNUSED (tty);
    NMH_UNUSED (utp);
#endif /* HAVE_GETUTXENT */

    exit (RCV_MOK);
}


static void
alrmser (int i)
{
    NMH_UNUSED (i);

    longjmp (myctx, 1);
}


static int
message_fd (char **vec)
{
    pid_t child_id;
    int bytes, seconds;
    int fd;
    char *tfile;
    struct stat st;

    if ((tfile = m_mktemp2(NULL, invo_name, &fd, NULL)) == NULL) {
	inform("unable to create temporary file in %s", get_temp_dir());
	return NOTOK;
    }
    (void) m_unlink(tfile);  /* Use fd, no longer need the file name. */

    if ((child_id = fork()) == NOTOK) {
	/* fork error */
	close (fd);
	return header_fd ();
    }
    if (child_id) {
	/* parent process */
	if (!setjmp (myctx)) {
	    SIGNAL (SIGALRM, alrmser);
	    bytes = fstat(fileno (stdin), &st) != NOTOK ? (int) st.st_size : 100;

	    /* amount of time to wait depends on message size */
	    if (bytes <= 100) {
		/* give at least 5 minutes */
		seconds = 300;
	    } else if (bytes >= 90000) {
		/* but 30 minutes should be long enough */
		seconds = 1800;
	    } else {
		seconds = (bytes / 60) + 300;
	    }
	    alarm ((unsigned int) seconds);
	    pidwait(child_id, OK);
	    alarm (0);

	    if (fstat (fd, &st) != NOTOK && st.st_size > 0)
		return fd;
	} else {
	    /*
	     * Ruthlessly kill the child and anything
	     * else in its process group.
	     */
	    killpg(child_id, SIGKILL);
	}
	close (fd);
	return header_fd ();
    }

    /* child process */
    rewind (stdin);
    if (dup2 (fd, 1) == NOTOK || dup2 (fd, 2) == NOTOK)
	_exit (-1);
    closefds (3);
    setpgid ((pid_t) 0, getpid ());	/* put in own process group */
    if (execvp (vec[0], vec) == NOTOK) {
        _exit (-1);
    }

    return NOTOK;
}


static int
header_fd (void)
{
    int fd;
    char *nfs;
    char *tfile = NULL;
    charstring_t scanl = NULL;

    if ((tfile = m_mktemp2(NULL, invo_name, &fd, NULL)) == NULL) {
	inform("unable to create temporary file in %s", get_temp_dir());
        return NOTOK;
    }
    (void) m_unlink(tfile);  /* Use fd, no longer need the file name. */

    rewind (stdin);

    /* get new format string */
    nfs = new_fs (form, format, SCANFMT);
    scan (stdin, 0, 0, nfs, width, 0, 0, NULL, 0L, 0, &scanl);
    scan_finished ();
    if (newline) {
	if (write (fd, "\n\r", 2) < 0) {
	    advise (tfile, "write LF/CR");
	}
    }
    if (write (fd, charstring_buffer (scanl), charstring_bytes (scanl)) < 0) {
	advise (tfile, "write");
    }
    charstring_free (scanl);
    if (bell) {
        if (write (fd, "\007", 1) < 0) {
	    advise (tfile, "write BEL");
        }
    }

    return fd;
}


#if HAVE_GETUTXENT
static void
alert (char *tty, int md)
{
    int i, td, mask;
    char buffer[BUFSIZ], ttyspec[BUFSIZ];
    struct stat st;

    snprintf (ttyspec, sizeof(ttyspec), "/dev/%s", tty);

    /*
     * The mask depends on whether we are checking for
     * write permission based on `biff' or `mesg'.
     */
    mask = biff ? S_IEXEC : (S_IWRITE >> 3);
    if (stat (ttyspec, &st) == NOTOK || (st.st_mode & mask) == 0)
	return;

    if (setjmp (myctx)) {
	alarm (0);
	return;
    }
    SIGNAL (SIGALRM, alrmser);
    alarm (2);
    td = open (ttyspec, O_WRONLY);
    alarm (0);
    if (td == NOTOK)
        return;

    lseek(md, 0, SEEK_SET);

    while ((i = read (md, buffer, sizeof(buffer))) > 0)
	if (write (td, buffer, i) != i)
	    break;

    close (td);
}
#endif /* HAVE_GETUTXENT */
